src/coprocessor/dag/expr/json.rs

```diff
@@ -156,59 +163,29 @@ impl<'a> JsonFuncArgsParser<'a> {
         JsonFuncArgsParser { ctx: ctx, row: row }
     }
 
-    #[inline]
-    fn parse<'b: 'a, It, T, F>(args: It, f: F) -> Result<Option<Vec<T>>>
-    where
-        It: Iterator<Item = &'b Expression>,
-        F: Fn(&'b Expression) -> Result<Option<T>>,
-    {
-        args.map(f).collect()
+    fn get_path_expr(&self, e: &Expression) -> Result<Option<PathExpression>> {
+        let s = try_opt!(e.eval_string_and_decode(self.ctx, self.row));
+        let expr = try!(parse_json_path_expr(&s));
+        Ok(Some(expr))
     }
 
-    fn get_path_exprs<'b: 'a, It>(&'a self, args: It) -> Result<Option<Vec<PathExpression>>>
-    where
-        It: Iterator<Item = &'b Expression>,
-    {
-        let func = |e: &'b Expression| {
-            let s = try_opt!(e.eval_string_and_decode(self.ctx, self.row));
-            parse_json_path_expr(&s).map_err(Error::from).map(Some)
-        };
-        JsonFuncArgsParser::parse(args, func)
+    fn get_path_exprs(&self, es: &[Expression]) -> Result<Option<Vec<PathExpression>>> {
+        es.iter().map(|e| self.get_path_expr(e)).collect()
     }
 
-    fn get_jsons<'b: 'a, It>(&'a self, args: It) -> Result<Option<Vec<Json>>>
-    where
-        It: Iterator<Item = &'b Expression>,
-    {
-        let func = |e: &'b Expression| {
-            let j = try!(e.eval_json(self.ctx, self.row))
-                .map(Cow::into_owned)
-                .unwrap_or(Json::None);
-            Ok(Some(j))
-        };
-        JsonFuncArgsParser::parse(args, func)
+    fn get_json(&self, e: &Expression) -> Result<Option<Json>> {
+        let j = try!(e.eval_json(self.ctx, self.row)).map_or(Json::None, Cow::into_owned);
+        Ok(Some(j))
     }
 
-    fn get_jsons_not_none<'b: 'a, It>(&'a self, args: It) -> Result<Option<Vec<Json>>>
-    where
-        It: Iterator<Item = &'b Expression>,
-    {
-        let func = |e: &'b Expression| {
-            let j = try_opt!(e.eval_json(self.ctx, self.row)).into_owned();
-            Ok(Some(j))
-        };
-        JsonFuncArgsParser::parse(args, func)
+    fn get_json_not_none(&self, e: &Expression) -> Result<Option<Json>> {
+        let j = try_opt!(e.eval_json(self.ctx, self.row)).into_owned();
+        Ok(Some(j))
     }
 
-    fn get_strings<'b: 'a, It>(&'a self, args: It) -> Result<Option<Vec<String>>>
-    where
-        It: Iterator<Item = &'b Expression>,
-    {
-        let func = |e: &'b Expression| {
-            let bytes = try_opt!(e.eval_string(self.ctx, self.row)).into_owned();
-            String::from_utf8(bytes).map(Some).map_err(Error::from)
-        };
-        JsonFuncArgsParser::parse(args, func)
+    fn get_string(&self, e: &Expression) -> Result<Option<String>> {
+        let bytes = try_opt!(e.eval_string(self.ctx, self.row)).into_owned();
```

Comment:
Is it a bug? Should we call `eval_string_and_decode` anywhere we need a `String`?